tls-codec/src/lib.rs

```diff
@@ -0,0 +1,80 @@
+use std::cell::Cell;
+
+mod arrays;
+#[cfg(feature = "hpke")]
+mod hpke;
+mod primitives;
+mod tls_vec;
+pub use tls_vec::{TlsVecU16, TlsVecU32, TlsVecU64, TlsVecU8};
+
+#[cfg(feature = "derive")]
+pub use tls_codec_derive::{TlsDeserialize, TlsSerialize};
+
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum Error {
+    EncodingError,
+    InvalidVectorLength,
+    InvalidInput,
+    DecodingError,
+}
+
+pub trait Serialize {
+    fn tls_serialize(&self, buffer: &mut Vec<u8>) -> Result<(), Error>;
+
+    fn serialize_detached(&self) -> Result<Vec<u8>, Error> {
+        let mut buffer = Vec::new();
+        self.tls_serialize(&mut buffer)?;
+        Ok(buffer)
+    }
+}
+
+pub trait Deserialize {
+    fn tls_deserialize(cursor: &Cursor) -> Result<Self, Error>
+    where
+        Self: Sized;
+
+    fn deserialize_detached(bytes: &[u8]) -> Result<Self, Error>
+    where
+        Self: Sized,
+    {
+        Self::tls_deserialize(&mut Cursor::new(bytes))
+    }
+}
+
+#[derive(Debug)]
+pub struct Cursor {
+    bytes: Vec<u8>,
+    position: Cell<usize>,
```

Comment:
I know this was also a problem before, but on 32bit platforms `usize` cannot index 64 bit vectors.

Reply:
Indeed, we should rework `Cursor` anyway later. I filed #319 to keep track of this specific issue.